ssl/s3_lib.c

```diff
@@ -4148,6 +4151,32 @@ const SSL_CIPHER *ssl3_choose_cipher(SSL *s, STACK_OF(SSL_CIPHER) *clnt,
     if (s->options & SSL_OP_CIPHER_SERVER_PREFERENCE || tls1_suiteb(s)) {
         prio = srvr;
         allow = clnt;
+#ifndef OPENSSL_NO_CHACHA
+        /* If ChaCha is at the top of the client preference list,
+           temporarily prioritize ChaCha ciphers in the servers list. */
+        if (s->options & SSL_OP_PRIORITIZE_CHACHA && sk_SSL_CIPHER_num(clnt) > 0) {
```

Comment:
One can wonder if it's appropriate to do so in tls1_suiteb case... Well, it might happen that Chacha won't make into server list in such case, but then making a copy of cipher list would be waste of resources...

Reply:
One would have to iterate through the stack to see if there's ChaCha to make the decision to re-order. But is that worth it?

Reply:
Here is my line of reasoning. Note that tls1_suiteb effectively takes precedence over SSL_OP_CIPHER_SERVER_PREFERENCE(*). It's natural to expect that it should take precedence even over SSL_OP_PRIORITIZE_CHACHA. So that for this reason alone one can wonder if check for SSL_OP_PRIORITIZE_CHACHA should be performed only in non-suiteb case. But on the other hand chances are that in suiteb case Chacha isn't even there(**) so one can use tls1_suiteb as criteria for not executing effectively no-op code.

(*) In sense that tls1_suiteb prioritizes server ciphers even without SSL_OP_CIPHER_SERVER_PREFERENCE. In the context one can even wonder if flipping the two in if condition would make code more readable. Or even `if (tls1_suiteb) {} else if (SSL_OP_CIPHER_SERVER_PREFERENCE) {} else {}`.

(**) This seems to be untrue in sense that if I s_server -suiteB_128 I can still connect with s_client -cipher ECDHE-RSA-CHACHA20-POLY1305. But nevertheless, first reason is probably sufficient by itself.

Reply:
Well, there is also chance that I'm missing something specific about tls1_suiteb. But it does *look* like that it would be appropriate to give precedence to tls1_suiteb even in SSL_OP_PRIORITIZE_CHACHA case.

Reply:
re: tls1_suiteb; I don't care, just give me a definitive answer. :)

Reply:
Among other things RFC6460 says that "A Suite B TLS server ... MUST accept ... the TLS_ECDHE_ECDSA_WITH_AES_* cipher suite *if* it is offered in the ClientHello message." Well, there still is certain ambiguity, because when it comes to client side it says that "One of these two cipher suites MUST be the first (most preferred) cipher suites in the ClientHello message." So that Chacha being first in list presented in ClientHello formally disqualifies it as compliant, and then one can wonder if server is in turn obliged to comply with the requirement... I'd say in case of doubt assume the worst, i.e. make server comply. Or in other words don't SSL_OP_PRIORITIZE_CHACHA in tls1_suiteb case.

Reply:
Done